xwiki-platform-core/xwiki-platform-security/xwiki-platform-security-requiredrights/xwiki-platform-security-requiredrights-api/src/main/java/org/xwiki/platform/security/requiredrights/RequiredRight.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.platform.security.requiredrights;
+
+import org.xwiki.model.EntityType;
+import org.xwiki.security.authorization.Right;
+import org.xwiki.stability.Unstable;
+import org.xwiki.text.XWikiToStringBuilder;
+
+/**
+ * Represents a required right for an entity, composed of a {@link Right}, an {@link EntityType} and a boolean optional
+ * field.
+ *
+ * @version $Id$
+ * @since 15.9RC1
+ */
+@Unstable
+public class RequiredRight
+{
+    private final Right right;
+
+    private final EntityType entityType;
+
+    private final boolean optional;
+
+    /**
+     * @param right the required right
+     * @param entityType the level at which the right is required (e.g., document, space, wiki)
+     * @param optional whether the right is optional or not, i.e., if the entity also works without the right or
+     *     not
+     */
+    public RequiredRight(Right right, EntityType entityType, boolean optional)
+    {
+        this.right = right;
+        this.entityType = entityType;
+        this.optional = optional;
+    }
+
+    /**
+     * @return the required right
+     */
+    public Right getRight()
+    {
+        return this.right;
+    }
+
+    /**
+     * @return the level at which the right is required (e.g., document, space, wiki)
+     */
+    public EntityType getEntityType()
+    {
+        return this.entityType;
+    }
+
+    /**
+     * @return whether the right is optional or not, i.e., if the entity also works without the right or not
```

Comment:
Why would we need a required right for an entity that would work without it? It could be useful to have a bit more doc / example for this.

Reply:
Yes, good point. I'll add it to the documentation, the use cases are as follows:

* A title that contains `#` or `$` (like "HashTag #XWiki") written by user without script right doesn't necessarily require script right but when saved by a user with script right, we cannot exclude that some code would be executed. So we say the right is optional. Note that there is also no way to escape this `#` or `$` that doesn't require script right, so there is absolutely no other way to write a title "HashTag #XWiki" if the user doesn't have script right.
* Any script might or might not need programming right, this is impossible to analyze. Even if there are checks for programming right in the code, the script might work just fine. In this case, we're also saying that programming right is optional.

Maybe "optional" is the wrong word, if you have a better idea let me know. I think the UI also needs to show these "optional" required rights differently.

Reply:
Ok so it feels more like "undecided" than "optional" as "impossible to decide whether or not the right is needed".

Reply:
or we want to be more explicit something like "needHumanAnalysis"

Reply:
I've renamed it to `manualReviewRequired` with a longer documentation. I hope this is clearer now.